gum/backend-windows/gumprocess-windows.c

```diff
@@ -305,13 +305,70 @@ _gum_process_enumerate_ranges (GumPageProtection prot,
   }
 }
 
+static gboolean
+gum_process_enumerate_heap_ranges (HANDLE heap, GumFoundMallocRangeFunc func,
+                                   gpointer user_data)
+{
+  PROCESS_HEAP_ENTRY entry;
+  GumMallocRangeDetails details;
+  GumMemoryRange gum_range;
+  gboolean carry_on, unlock_heap;
+
+  // HeapLock may fail but it doesn't seem to have any real consequences...
+  unlock_heap = HeapLock (heap);
+
+  details.range = &gum_range;
+  carry_on = TRUE;
+  entry.lpData = NULL;
+  while (carry_on && HeapWalk (heap, &entry)) {
+    if ((entry.wFlags & PROCESS_HEAP_ENTRY_BUSY) != 0) {
+      gum_range.base_address = entry.lpData;
+      gum_range.size = entry.cbData;
+      carry_on = func (&details, user_data);
+    }
+  }
+  if (unlock_heap)
+    HeapUnlock (heap);
+
+  return carry_on;
+}
+
 void
 gum_process_enumerate_malloc_ranges (GumFoundMallocRangeFunc func,
                                      gpointer user_data)
 {
-  /* Not implemented */
-  (void) func;
-  (void) user_data;
+  DWORD nb_heaps, i;
+  HANDLE process_heap;
+  HANDLE * heaps;
+
+  process_heap = GetProcessHeap ();
+  if (!gum_process_enumerate_heap_ranges (process_heap, func, user_data))
+    return;
+
+  while (TRUE) {
+    DWORD new_nb_heaps;
+    nb_heaps = GetProcessHeaps (0, NULL);
+    if (!nb_heaps)
+      return;
+    heaps = HeapAlloc (process_heap, 0, nb_heaps * sizeof(HANDLE));
+    if (!heaps)
+      return;
+    new_nb_heaps = GetProcessHeaps (0, NULL);
+    if (!new_nb_heaps) {
+      HeapFree (process_heap, 0, heaps);
+      return;
+    }
+    if (new_nb_heaps == nb_heaps)
+      break;
+    HeapFree (process_heap, 0, heaps);
+  }
```

Comment:
Let's add a blank line here; the two loops represent different logical operations.